### PR TITLE
Automate uploads and amendments

### PR DESCRIPTION
**Pull request by hancush**
## Description

This PR:

- Containerizes data making
- Migrates to our contemporary strategy for running migrations in the entrypoint
- Unifies the upload and amendment data pipelines
- Adds gut checks to intermediate targets where data can be lost
- Adds a management command to facilitate uploading new data or amending existing data
- Adds targets to upload or amend a given file using the management command
- Fixes two import bugs
  - Employer names can be changed via the admin interface, but a corresponding alias is not created for the new name. If the updated name appears in a new upload, the import tries to insert it as a new employer, violating the unique name constraint. This PR adds a data migration to insert aliases based on Django admin log entries indicating a name change, and adds logic to the employer admin to create aliases on name change automatically, moving forward.
  - When adding a third year, our query to link people failed when a person already had two salaries, because having more than one salary caused a person to appear twice in our query selecting existing people, and we only link if there is one matching person. I added a distinct on person ID to the query to resolve this issue. 
 
Connects #248.

### Testing instructions

_Tested locally._

### CODE REVIEW

File: data/payroll.mk [hunk 1]
```diff
@@ -0,0 +1,69 @@
+.INTERMEDIATE : %-with-agencies.csv %-with-data-year.csv \
```
Comment (hancush) [inline data/payroll.mk:1]: Almost all of this is copied from `data/amended_files.mk`. The new things are the `import/` and `amend/` targets, and the gut checks in the `with-agencies` and `no-salary-omitted` targets.

**Comment by hancush** [comment]
@fgregg This is ready for preliminary review. I would especially appreciate feedback on overall approach and thoughts on high value tests to write, if any.

File: Dockerfile [hunk 2]
```diff
@@ -1,37 +1,13 @@
-# Extend the base Python image
```
Comment (fgregg) [inline Dockerfile:1]: why the change to this dockerfile?

File: data/payroll.mk [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+	perl -pe "s/$$/,$$(cut -d '-' -f 1 <<< $*)/" $< > $@
+
+%-leading-whitespace-trimmed.csv : %-no-salary-omitted.csv
+	perl -pe 's/^\s+//' $< > $@
+
+%-no-salary-omitted.csv : %-with-valid-start-dates.csv
+	# Remove records where no salary was reported and validate that the output
+	# file has the expected number of lines.
+	csvgrep -c base_salary,extra_pay -r '^$$' -i $< > $@; \
+	N_REMOVED=$$(csvgrep -c base_salary,extra_pay -r '^$$' $< | tail -n +2 | wc -l | xargs); \
```
Comment (fgregg) [inline data/payroll.mk:32]: why do you need xargs here?

File: data/payroll.mk [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+	echo "Removed $$N_REMOVED records without salary"; \
+	EXPECTED_LINES=$$(expr $$(grep . $< | wc -l | grep -Eo '[0-9]+' | head -n 1) - $$N_REMOVED); \
+	OUTFILE_LINES=$$(grep . $@ | wc -l | grep -Eo '[0-9]+' | head -n 1); \
```
Comment (fgregg) [inline data/payroll.mk:35]: "csvstat --count" would also be easier to follow

File: data/payroll.mk [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+	fi
+
+%-with-valid-start-dates.csv : %-with-agencies.csv
+	# Remove invalid dates.
+	cat $< | python data/processors/validate_dates.py > $@
+
+%-with-agencies.csv : data/raw/payroll-actual-%.csv data/raw/foia-source-lookup.csv
+	# Join standard data with agency lookup and validate that the output file
+	# has the expected number of lines.
+	csvjoin -c id,ID -e IBM852 $^ > $@ && \
```
Comment (fgregg) [inline data/payroll.mk:50]: why line continuations?

File: data/payroll.mk [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+	INFILE_LINES=$$(grep . $< | wc -l | grep -Eo '[0-9]+' | head -n 1); \
```
Comment (fgregg) [inline data/payroll.mk:51]: ditto csvstat

File: data/payroll.mk [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+	OUTFILE_LINES=$$(grep . $@ | wc -l | grep -Eo '[0-9]+' | head -n 1); \
```
Comment (fgregg) [inline data/payroll.mk:52]: ditto csvstat

File: data/payroll.mk [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+		make $*-missing-agencies.csv; \
+	fi
+
+%-missing-agencies.csv : data/raw/payroll-actual-%.csv data/raw/foia-source-lookup.csv
+	# If this recipe fires, one or more responding agencies cited in incoming
+	# data are missing from the FOIA source lookup. The output contains their
+	# id and employer name. Either the lookup table needs to be updated, or the
+	# data needs to be amended to reference an existing code.
+	csvjoin --left -c id,ID -e IBM852 $^ | \
+	csvgrep -c Employer -r "^$$" | \
```
Comment (fgregg) [inline data/payroll.mk:66]: please indent csvgrep and the following lines to make it a bit clearer that this is one long chain.

File: data_import/tasks.py [hunk 9]
```diff
@@ -36,9 +37,18 @@ class and extend common functionality via super() in the
         '''
         from data_import.models import StandardizedFile
 
-        self.s_file = StandardizedFile.objects.get(id=s_file_id)
+        print('Setup fired')
```
Comment (fgregg) [inline data_import/tasks.py:40]: let's use django/python logging mechanisms

File: data_import/utils/csv_meta.py [hunk 10]
```diff
@@ -72,32 +69,27 @@ def trim_extra_fields(self):
...
-            with open(outfile_name, 'w', encoding='utf-8') as outfile:
-                writer = csv.DictWriter(outfile, fieldnames=self.REQUIRED_FIELDS)
-                writer.writeheader()
+        outfile_name = '/tmp/{}'.format(basename(self.file.name))
 
-                for row in reader:
-                    out_row = {field: row[field] for field in self.REQUIRED_FIELDS}
-                    writer.writerow(out_row)
+        with open(outfile_name, 'w', encoding='utf-8') as outfile:
+            writer = csv.DictWriter(outfile, fieldnames=self.REQUIRED_FIELDS)
```
Comment (fgregg) [inline data_import/utils/csv_meta.py:86]: nice!

File: payroll/management/commands/import_data.py [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+            'host': django_conn.get('host', ''),
+            'port': django_conn.get('port', ''),
+            'database': django_conn.get('database', ''),
+        }
+
+        self.engine = sa.create_engine(URL('postgresql', **conn_kwargs))
+
+        self.upload()
+
+    def prompt(self, prompt):
```
Comment (fgregg) [inline payroll/management/commands/import_data.py:58]: maybe "bail_out_prompt" would be clearer

**Review by fgregg — DISMISSED** [review]
why the changes to dockerfile?

**Comment by hancush** [comment]
# TODO

- [ ] Set `CELERY_RESULT_BACKEND = REDIS_URL` in deployment settings
- [ ] Debug failures to upload to S3. Getting...

```
# info about PUT request
encountered a retry exception: Saw RequestTimeout, retrying
establishing HTTPS connection: host=bga-payroll-database.s3.amazonaws.com, kwargs={'timeout': 70, 'port': 443}
```

File: .travis.yml [hunk 12]
```diff
@@ -25,7 +25,7 @@ before_script:
 - cp configs/local_settings.py.travis bga_database/local_settings.py
 
 script:
-- pytest --reuse-db
+- pytest -sv
```
Comment (hancush) [inline .travis.yml:28]: `--reuse-db` doesn't work for CI.

File: data/payroll.mk [hunk 13]
```diff
@@ -0,0 +1,68 @@
+.INTERMEDIATE : %-with-agencies.csv %-with-data-year.csv \
+	%-with-valid-start-dates.csv %-salary-summed.csv \
+	%-no-salary-omitted.csv %-with-agencies.csv
+
+.PRECIOUS : %-with-agencies.csv data/output/%
+
+define get_record_count
+$$(csvstat --count $(1) | cut -d ' ' -f3)
```
Comment (hancush) [inline data/payroll.mk:8]: Nice legibility improvement. Thanks for the suggestion, @fgregg.

File: data_import/utils/import_utility.py [hunk 14]
```diff
@@ -480,8 +480,10 @@ def select_raw_person(self):
               LEFT JOIN payroll_employeralias AS department_alias
                 ON child.id = department_alias.employer_id
             ), existing AS (
-              /* Create a view of existing people and their employers. */
-              SELECT
+              /* Create a view of existing people and their employers. Select
```
Comment (hancush) [inline data_import/utils/import_utility.py:483]: Linking fix, as discussed in the PR description.

File: payroll/admin.py [hunk 15]
```diff
@@ -17,7 +18,20 @@ class AdminEmployer(admin.ModelAdmin):
     readonly_fields = ('slug', 'parent', 'vintage',)
 
     def save_model(self, request, obj, form, change):
-        super().save_model(request, obj, form, change)
+        with transaction.atomic():
+            super().save_model(request, obj, form, change)
+
+            if change:
+                # Get or create an alias for the given employer. Don't filter
```
Comment (hancush) [inline payroll/admin.py:25]: Alias fix 1/2, as discussed in the PR description.

File: payroll/migrations/0034_reflect_aliases.py [hunk 16]
```diff
@@ -0,0 +1,48 @@
+# Generated by Django 2.2.9 on 2021-08-30 18:47
+'''
+Create aliases for employer name changes in the admin interface.
```
Comment (hancush) [inline payroll/migrations/0034_reflect_aliases.py:3]: Alias fix 2/2, as discussed in the PR description.

File: payroll/models.py [hunk 17]
```diff
@@ -229,17 +229,17 @@ class EmployerAlias(AliasModel):
...
 
-        if self.employer.is_department:
-            duplicate_alias = type(self).objects.filter(
-                Q(name=self.name) & Q(employer__parent_id=self.employer.parent.id)
-            )
-        else:
-            duplicate_alias = type(self).objects.filter(
-                Q(name=self.name) & Q(employer__parent_id__isnull=True)
-            )
+        existing_alias = type(self).objects.filter(
```
Comment (hancush) [inline payroll/models.py:232]: Refinement of existing logic. Aliases are always associated with the employer ID, so we can use the same logic for units and departments.

File: payroll/models.py [hunk 18]
```diff
@@ -229,17 +229,17 @@ class EmployerAlias(AliasModel):
...
+            name=self.name,
+            employer=self.employer
+        ).exclude(
+            id=self.id
+        )
```
Comment (hancush) [inline payroll/models.py:237]: This prevents creating new aliases that are duplicates, while providing for updates to an existing alias.

File: data/payroll.mk [hunk 19]
```diff
@@ -0,0 +1,68 @@
...
+	cat $< | python data/processors/validate_dates.py > $@
+
+%-with-agencies.csv : %-whitespace-trimmed.csv data/raw/foia-source-lookup.csv
+	# Join standard data with agency lookup and validate that the output file
+	# has the expected number of lines.
+	csvjoin -c id,ID -e IBM852 $^ > $@
+	if [ $(call get_record_count,$<) -eq $(call get_record_count,$@) ]; then \
+		echo "No lines lost in join. Proceeding..."; \
+	else \
+		echo "$< has $$INFILE_LINES lines, but $@ has $$OUTFILE_LINES lines. Generating missing agencies..."; \
```
Comment (fgregg) [inline data/payroll.mk:48]: how does $INFILE_LINES and $OUTFILE_LINES get set?

Comment (hancush) [reply data/payroll.mk:50]: @fgregg Whoops, fixed in https://github.com/datamade/bga-payroll/pull/526/commits/10058ca9bbaaf91bcd6455f8ae909277ba5230a1. Need the line continuations in this recipe because the Make runs each command in its own subshell, so they need to be part of the same command as the conditional that references them.

Comment (fgregg) [reply data/payroll.mk:50]: word
